hibernate-core/src/main/java/org/hibernate/query/sqm/tree/domain/SqmMapJoin.java

```diff
@@ -76,75 +75,81 @@ public SqmMapJoin(
 		final SqmPathSource keyPathSource = getReferencedPathSource().getKeyPathSource();
```

Comment:
Ideally, the method would just use `org.hibernate.query.sqm.tree.domain.AbstractSqmPath#resolvePath(java.lang.String, org.hibernate.query.sqm.SqmPathSource<?>)`, but we need a special path source implementation for the key type as per the todo in `org.hibernate.query.sqm.internal.SqmMappingModelHelper#resolveSqmKeyPathSource`. Apart from that, it would be nice if you would reuse the same approach for `org.hibernate.query.hql.internal.SemanticQueryBuilder#visitCollectionElementNavigablePath`, `org.hibernate.query.hql.internal.SemanticQueryBuilder#visitCollectionIndexFunction` and `org.hibernate.query.hql.internal.SemanticQueryBuilder#visitMapKeyNavigablePath`. In fact, every `createSqmPath` call in `SemanticQueryBuilder` is questionable IMO as that should all be handled through `resolvePath` or some other path related methods.